File: cli/src/commands/notes.ts
import { Command } from "commander";
import { createClient, getGlobalOpts } from "../client.js";
import { output } from "../output.js";
import { withErrorHandler } from "../errors.js";

function registerNotesAccessCommands(parent: Command): void {
  const access = parent
    .command("access")
    .description("Per-note access grants");

  access
    .command("list <note-id>")
    .description("List grants on a note")
    .action(
      withErrorHandler(async function (this: Command, noteId: string) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const rows = await inkbox.notes.access.list(noteId);
        output(rows, {
          json: !!opts.json,
          columns: ["id", "noteId", "identityId", "createdAt"],
        });
      }),
    );

  access
    .command("grant <note-id> <identity-id>")
    .description("Grant a specific identity access (admin + JWT only)")
    .action(
      withErrorHandler(async function (
        this: Command,
        noteId: string,
        identityId: string,
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const grant = await inkbox.notes.access.grant(noteId, identityId);
        output(grant as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  access
    .command("revoke <note-id> <identity-id>")
    .description("Revoke a specific identity's grant (self-only for agents)")
    .action(
      withErrorHandler(async function (
        this: Command,
        noteId: string,
        identityId: string,
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        await inkbox.notes.access.revoke(noteId, identityId);
        console.log(`Revoked identity ${identityId} on note ${noteId}.`);
      }),
    );
}

export function registerNotesCommands(program: Command): void {
  const notes = program
    .command("notes")
    .description("Org-scoped notes with per-identity grants");

  notes
    .command("list")
    .description("List accessible notes")
    .option("--q <query>", "Substring search (≤200 chars)")
    .option("--identity <uuid>", "Filter to notes visible to an identity")
    .option("--order <order>", "recent (default) or created")
    .option("--limit <n>", "Max rows (1–200, default 50)", (v) => parseInt(v, 10))
    .option("--offset <n>", "Offset", (v) => parseInt(v, 10))
    .action(
      withErrorHandler(async function (
        this: Command,
        cmdOpts: {
          q?: string;
          identity?: string;
          order?: string;
          limit?: number;
          offset?: number;
        },
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const rows = await inkbox.notes.list({
          q: cmdOpts.q,
          identityId: cmdOpts.identity,
          order: cmdOpts.order,
          limit: cmdOpts.limit,
          offset: cmdOpts.offset,
        });
        output(rows, {
          json: !!opts.json,
          columns: ["id", "title", "createdBy", "status", "createdAt", "updatedAt"],
        });
      }),
    );

  notes
    .command("get <note-id>")
    .description("Fetch a single note with inlined grants")
    .action(
      withErrorHandler(async function (this: Command, noteId: string) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const note = await inkbox.notes.get(noteId);
        output(note as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  notes
    .command("create")
    .description("Create a note")
    .requiredOption("--body <text>", "Body text (1–100 000 chars, required)")
    .option("--title <text>", "Optional title (≤200 chars)")
    .action(
      withErrorHandler(async function (
        this: Command,
        cmdOpts: { body: string; title?: string },
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const note = await inkbox.notes.create({
          body: cmdOpts.body,
          title: cmdOpts.title,
        });
        output(note as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  notes
    .command("update <note-id>")
    .description("Update title and/or body")
    .option("--body <text>", "New body text")
    .option("--title <text>", 'New title (pass "" to clear)')
    .action(
      withErrorHandler(async function (
        this: Command,
        noteId: string,
        cmdOpts: { body?: string; title?: string },
      ) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        const patch: { body?: string; title?: string | null } = {};
        if (cmdOpts.body !== undefined) patch.body = cmdOpts.body;
        if (cmdOpts.title !== undefined) {
          patch.title = cmdOpts.title === "" ? null : cmdOpts.title;
        }
        const note = await inkbox.notes.update(noteId, patch);
        output(note as unknown as Record<string, unknown>, { json: !!opts.json });
      }),
    );

  notes
    .command("delete <note-id>")
    .description("Delete a note")
    .action(
      withErrorHandler(async function (this: Command, noteId: string) {
        const opts = getGlobalOpts(this);
        const inkbox = createClient(opts);
        await inkbox.notes.delete(noteId);
        console.log(`Deleted note ${noteId}.`);
      }),
    );

  registerNotesAccessCommands(notes);
}

File: sdk/python/inkbox/notes/resources/notes.py
"""
inkbox/notes/resources/notes.py

Notes CRUD + per-note access subresource.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal
from uuid import UUID

from inkbox.notes.resources.note_access import NoteAccessResource
from inkbox.notes.types import Note

if TYPE_CHECKING:
    from inkbox._http import HttpTransport

_BASE = "/notes"
_UNSET = object()


class NotesResource:
    """Org-scoped notes with per-identity access grants."""

    def __init__(self, http: HttpTransport) -> None:
        self._http = http
        self._access = NoteAccessResource(http)

    @property
    def access(self) -> NoteAccessResource:
        return self._access

    def list(
        self,
        *,
        q: str | None = None,
        identity_id: UUID | str | None = None,
        limit: int | None = None,
        offset: int | None = None,
        order: Literal["recent", "created"] | str | None = None,
    ) -> list[Note]:
        """List accessible notes.

        Args:
            q: Substring search (≤200 chars).
            identity_id: Filter to notes visible to a specific identity.
            limit: 1–200 (server default 50).
            offset: Offset for paging.
            order: ``"recent"`` (default) or ``"created"``.
        """
        params: dict[str, Any] = {}
        if q is not None:
            params["q"] = q
        if identity_id is not None:
            params["identity_id"] = str(identity_id)
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        if order is not None:
            params["order"] = order
        data = self._http.get(_BASE, params=params)
        items = data["items"] if isinstance(data, dict) and "items" in data else data
        return [Note._from_dict(n) for n in items]

    def get(self, note_id: UUID | str) -> Note:
        data = self._http.get(f"{_BASE}/{note_id}")
        return Note._from_dict(data)

    def create(self, *, body: str, title: str | None = None) -> Note:
        """Create a note.

        Agent-created notes auto-grant the creator. Human-created notes start
        with zero grants and are invisible to all agents until granted.
        """
        payload: dict[str, Any] = {"body": body}
        if title is not None:
            payload["title"] = title
        data = self._http.post(_BASE, json=payload)
        return Note._from_dict(data)

    def update(
        self,
        note_id: UUID | str,
        *,
        title: str | None = _UNSET,  # type: ignore[assignment]
        body: str = _UNSET,  # type: ignore[assignment]
    ) -> Note:
        """JSON-merge-patch update.

        ``title=None`` clears the title column (200 OK). ``body=None`` is
        **not** a legal operation — the body column is required; the server
        returns 422 if you try.
        """
        payload: dict[str, Any] = {}
        if title is not _UNSET:
            payload["title"] = title
        if body is not _UNSET:
            payload["body"] = body
        data = self._http.patch(f"{_BASE}/{note_id}", json=payload)
        return Note._from_dict(data)

    def delete(self, note_id: UUID | str) -> None:
        self._http.delete(f"{_BASE}/{note_id}")
